### PR TITLE
Don't throw exception, instead user should find out at compile time that they're passing the wrong argument

### PR DESCRIPTION
**Pull request by premnirmal**
Inside AppObservable, you have a method `bindFragment(Object,Observable<T>)`. Instead of accepting an `Object` parameter, you should have two methods, one accepting a `Fragment` and the other accepting a `android.support.v4.app.Fragment`.

### CODE REVIEW

**Comment by mttkay** [comment]
I've had the same discussion with @JakeWharton before. I'm still fairly sure I did this because if the support library is not on the classpath, this will crash at runtime when the class is loaded. Jake though otherwise. Not sure if you verified that this will actually work when you do not link the support library?

**Comment by premnirmal** [comment]
@mttkay Isn't the support library already int he classpath https://github.com/ReactiveX/RxAndroid/blob/0.x/build.gradle#L26 ?

File: rxandroid/src/main/java/rx/android/app/AppObservable.java [hunk 1]
```diff
@@ -62,10 +61,10 @@ public Boolean call(android.support.v4.app.Fragment fragment) {
 
     /**
      * Binds the given source sequence to an activity.
-     * <p>
+     * <p/>
```
Comment (JakeWharton) [inline rxandroid/src/main/java/rx/android/app/AppObservable.java]: This is wrong, revert. Paragraphs are not self-closing and doclint on JDK 8 will fail the build for this.

File: rxandroid/src/main/java/rx/android/app/AppObservable.java [hunk 2]
```diff
@@ -79,13 +78,30 @@ public Boolean call(android.support.v4.app.Fragment fragment) {
         return source.observeOn(mainThread()).lift(new OperatorConditionalBinding<T, Activity>(activity, ACTIVITY_VALIDATOR));
     }
 
+    @Deprecated
+    public static <T> Observable<T> bindFragment(Object fragment, Observable<T> source) {
+        return bindFragmentInternal(fragment, source);
+    }
+
+    private static <T> Observable<T> bindFragmentInternal(Object fragment, Observable<T> source) {
```
Comment (JakeWharton) [inline rxandroid/src/main/java/rx/android/app/AppObservable.java]: Destroy this method and move the respective implementations to the appropriate methods.

File: rxandroid/src/main/java/rx/android/app/AppObservable.java [hunk 3]
```diff
@@ -79,13 +78,30 @@ public Boolean call(android.support.v4.app.Fragment fragment) {
         return source.observeOn(mainThread()).lift(new OperatorConditionalBinding<T, Activity>(activity, ACTIVITY_VALIDATOR));
     }
 
+    @Deprecated
+    public static <T> Observable<T> bindFragment(Object fragment, Observable<T> source) {
```
Comment (JakeWharton) [inline rxandroid/src/main/java/rx/android/app/AppObservable.java]: Just delete. The overloads will resolve on upgrade for anyone using this correctly. And if you weren't then we just turned your runtime failure to a compile time failure.

**Comment by JakeWharton** [comment]
This will crash on Android 2.1 or earlier only.

**Comment by JakeWharton** [comment]
Easy to verify: just create a new empty project without the support library and call `bindFragment(android.app.Fragment)` and run it on any new OS.

**Comment by mttkay** [comment]
>  Isn't the support library already in the classpath https://github.com/ReactiveX/RxAndroid/blob/0.x/build.gradle#L26 ?

It's linked with `provided` scope. That means, it's in the classpath visible to the respective compilation tasks in Gradle, but it will not be available at runtime for any clients depending on the RxJava library. The way I read this is: if the class loader loads this class, and finds a reference to a symbol from the (missing) support library, it will shake its fist at you and crash.

If, however, the method signature declares `Object`, and we do the check/cast lazily upon method invocation, we can prevent this from happening since the code path will never execute unless the support Fragments are available. Does that make sense?

**Comment by JakeWharton** [comment]
> if the class loader loads this class, and finds a reference to a symbol from the (missing) support library, it will shake its fist at you and crash.

This is the behavior that changed with Android 2.2 and is no longer true.

**Comment by mttkay** [comment]
That's not to say that I'd love to see code like this go. I guess I just couldn't come up with a better solution when I first wrote this, so by all means please rewrite it. Just do it in a way that doesn't force support-v* on people.

**Comment by mttkay** [comment]
> This is the behavior that changed with Android 2.2 and is no longer true.

Right, you mentioned that before. Fair enough then.

File: rxandroid/src/main/java/rx/android/app/AppObservable.java (hunk 2 above)
Comment (premnirmal) [inline rxandroid/src/main/java/rx/android/app/AppObservable.java]: done, thanks

**Comment by premnirmal** [comment]
> if the class loader loads this class, and finds a reference to a symbol from the (missing) support library, it will shake its fist at you and crash.

> > This is the behavior that changed with Android 2.2 and is no longer true.

@mttkay @JakeWharton , yes you're right. It won't crash, but it won't compile because of the method signature has the support fragment, which isn't included in the classpath.

> Easy to verify: just create a new empty project without the support library and call bindFragment(android.app.Fragment) and run it on any new OS.
Tried it, won't compile because of the `bindFragment(android.support.v4.app.Fragment)` method.

Thanks for confirming.

Closing now.

**Comment by JakeWharton** [comment]
It's already compiled in the class file. The problem (or rather, lack of one) would be in the source that was trying compile against this class file and link against its methods.

**Comment by premnirmal** [comment]
yeah, the sample app is what doesn't compile

**Comment by JakeWharton** [comment]
With what error?

**Comment by premnirmal** [comment]
`Error:(77, 41) error: cannot access Fragment`
`class file for android.support.v4.app.Fragment not found`

on the line where I call `AppObservable.bindFragment(Fragment)`

**Comment by premnirmal** [comment]
here https://github.com/ReactiveX/RxAndroid/blob/0.x/sample-app/src/main/java/rx/android/samples/ListeningFragmentActivity.java#L77

**Comment by JakeWharton** [comment]
Interesting. Will try. We definitely do this exact thing without problems elsewhere.

**Comment by JakeWharton** [comment]
Forcing conversation back to #95.

**Pull request closed without merging**